pwiz_tools/Skyline/EditUI/AssociateProteinsDlg.cs

```diff
@@ -178,36 +190,84 @@ public void UseFastaFile(string file)
         }
 
         // creates dictionary of a protein to a list of peptides that match
-        private void FindProteinMatchesWithFasta(Stream fastaFile)
+        private List<KeyValuePair<FastaSequence, List<IdentityPath>>> FindProteinMatchesWithFasta(Stream fastaFile)
         {
-            var proteinAssociations = new List<KeyValuePair<FastaSequence, List<PeptideDocNode>>>();
             var peptidesForMatching = ListPeptidesForMatching();
-            using (var reader = new StreamReader(fastaFile))
+            if (peptidesForMatching.Count == 0)
+            {
+                return new List<KeyValuePair<FastaSequence, List<IdentityPath>>>();
+            }
+
+            int maxLength = peptidesForMatching.Max(peptide => GetPeptideSequence(peptide).Length);
+            long streamLength = fastaFile.Length;
+            var proteinAssociations = new List<Tuple<int, FastaSequence, List<IdentityPath>>>();
+            using (var longWaitDlg = new LongWaitDlg())
             {
-                foreach (var seq in FastaData.ParseFastaFile(reader))
+                longWaitDlg.Message = Resources.AssociateProteinsDlg_FindProteinMatchesWithFasta_Finding_peptides_in_FASTA_file;
+                longWaitDlg.PerformWork(this, 1000, broker =>
                 {
-                    var fasta = new FastaSequence(seq.Name, null, null, seq.Sequence);
-                    var matches = new List<PeptideDocNode>();
-                    foreach (var peptide in peptidesForMatching)
+                    var fastaRecords = ParseFastaFileWithProgress(fastaFile).Select((tuple, index) =>
+                        Tuple.Create(index, tuple.Item1, tuple.Item2));
+                    ParallelEx.ForEach(fastaRecords, recordTuple =>
                     {
-                        // TODO(yuval): does digest matter?
-                        if (fasta.Sequence.IndexOf(peptide.Peptide.Target.Sequence, StringComparison.Ordinal) < 0)
+                        int index = recordTuple.Item1;
+                        int progressValue = (int) (recordTuple.Item2 * 100 / streamLength);
+                        var seq = recordTuple.Item3;
+                        var fasta = new FastaSequence(seq.Name, null, null, seq.Sequence);
+                        var substringFinder = new SubstringFinder(seq.Sequence, maxLength);
+                        var matches = new List<IdentityPath>();
+                        foreach (var peptide in peptidesForMatching)
+                        {
+                            // TODO(yuval): does digest matter?
+                            if (!substringFinder.ContainsSubstring(GetPeptideSequence(peptide)))
+                            {
+                                continue;
+                            }
+
+                            matches.Add(peptide);
+                        }
+
+                        lock (proteinAssociations)

```

Comment:
I prefer not locking in these and instead preallocating an array the size of your fastaRecords. Inside the loop just assign your matches to that array without locking. When the loop finishes you can use a single-threaded loop through that array to build this list.